### PR TITLE
Added /interactive command to demo interactive button functionality

### PR DESCRIPTION
**Pull request by rfoyard**
Related to mattermost/mattermost-server#12697

This is an example to demo the usage of the interactive buttons.  
- Adds an `/interactive` command that display an interactive button
- When the button is clicked, a new post is created containing the request's json.

### CODE REVIEW

File: server/README.md [hunk 1]
```diff
@@ -69,15 +68,17 @@ The `/crash` command demonstrates crashing the plugin (and the server recovering
 
 The `/dialog` command demonstrates [Interactive Dialogs](https://docs.mattermost.com/developer/interactive-dialogs.html). Use `/dialog help` for its usage in this demo plugin.
 
-## [http\_hooks.go](http_hooks.go)
+The `/interactive` command demonstrates the usage of interactive buttons.
```
Comment (hanzei) [inline server/README.md]:
```suggestion
The `/interactive` command demonstrates the usage of interactive message buttons.
```

File: server/command_hooks.go [hunk 2]
```diff
@@ -82,6 +83,15 @@ func (p *Plugin) registerCommands() error {
 		return errors.Wrapf(err, "failed to register %s command", commandTriggerDialog)
 	}
 
+	if err := p.API.RegisterCommand(&model.Command{
+		Trigger:          commandTriggerInteractive,
+		AutoComplete:     true,
+		AutoCompleteHint: "",
+		AutoCompleteDesc: "Interactive button demonstration",
```
Comment (hanzei) [inline server/command_hooks.go]:
```suggestion
		AutoCompleteDesc: "Demonstrates  interactive message buttons.",
```

File: server/command_hooks.go [hunk 3]
```diff
@@ -405,6 +417,39 @@ func getDialogWithIntroductionText(introductionText string) model.Dialog {
...
+
+	post := &model.Post{
+		ChannelId: args.ChannelId,
+		UserId:    p.botId,
+		Message:   "Test interactive button",
+		Props: model.StringInterface{
+			"attachments": []*model.SlackAttachment{{
+				Actions: []*model.PostAction{{
+					Integration: &model.PostActionIntegration{
+						URL: fmt.Sprintf("%s/plugins/%s/interaction/action", siteURL, manifest.Id),
```
Comment (hanzei) [inline server/command_hooks.go]: The `siteURL` is actually not longer needed.

```suggestion
						URL: fmt.Sprintf("/plugins/%s/interaction/action", manifest.Id),
```

Comment (hanzei) [inline server/command_hooks.go]: Maybe went should use a more specific URL here. Something like e.g. `interactive/button/1`.

File: server/http_hooks.go [hunk 4]
```diff
@@ -229,6 +231,37 @@ func (p *Plugin) handleEphemeralDelete(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	user, appErr := p.API.GetUser(request.UserId)
+	if appErr != nil {
+		p.API.LogError("failed to get user for interaction action", "err", appErr.Error())
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	requestJSON, _ := json.MarshalIndent(request, "", "  ")
```
Comment (hanzei) [inline server/http_hooks.go]: Please check the error.

Comment (rfoyard) [reply server/http_hooks.go]: Oops, copied this from line 54 and did not modify thereafter, will do.

**Comment by rfoyard** [comment]
Thank you @hanzei

**Comment by hanzei** [comment]
Thanks _you_ for the PR!

**Comment by prapti** [comment]
@rfoyard Could you please add some test steps in the description?
@hanzei There is no Jira ticket for this particular PR but would https://mattermost.atlassian.net/browse/MM-10164 serve as the ticket for this PR as well?

**Comment by hanzei** [comment]
@prapti Yes, we can use this JIRA ticket.

**Comment by prapti** [comment]
@rfoyard Thank you!
A couple questions:
1. Clicking on the button generates a response that is grouped together in the reply thread (this is great!) but the reply count on the reply button does not get updated.
2. Issuing a `/interactive` command on the RHS reply thread generates the button in the main center channel. Is this by design? Would it be possible to generate the button on the RHS?
Since the interactive button itself does what is expected, we could merge this one and open improvement tickets if the issues are out of scope for this PR.
@hanzei @marianunez are there any thoughts/comments on these questions?

**Comment by rfoyard** [comment]
> 1. Clicking on the button generates a response that is grouped together in the reply thread (this is great!) but the reply count on the reply button does not get updated.

This is the same behaviour as the `/dialog` command.  
I am not sure how the ReplyCount is updated nor if it's even possible for bots' posts.

> 2. Issuing a `/interactive` command on the RHS reply thread generates the button in the main center channel. Is this by design? Would it be possible to generate the button on the RHS?

For this one, it's the same behaviour as the `/ephemeral` command.  
I just tried to create the button in the RHS, but I could not get the response to be grouped together (neither in the center channel nor in the RHS)

File: server/command_hooks.go [hunk 5]
```diff
@@ -405,6 +417,38 @@ func getDialogWithIntroductionText(introductionText string) model.Dialog {
 	return dialog
 }
 
+func (p *Plugin) executeCommandInteractive(args *model.CommandArgs) *model.CommandResponse {
+
+	post := &model.Post{
+		ChannelId: args.ChannelId,
+		UserId:    p.botId,
```
Comment (hanzei) [inline server/command_hooks.go]: Please add to fix the second issue
```suggestion
		RootId:    args.RootId,
		UserId:    p.botId,
```

File: server/command_hooks.go [hunk 6]
```diff
@@ -405,6 +417,38 @@ func getDialogWithIntroductionText(introductionText string) model.Dialog {
 	return dialog
 }
 
+func (p *Plugin) executeCommandInteractive(args *model.CommandArgs) *model.CommandResponse {
+
+	post := &model.Post{
+		ChannelId: args.ChannelId,
```
Comment (hanzei) [inline server/command_hooks.go]: Please add this to fix the second issue
```suggestion
		ChannelId: args.ChannelId,
		RootId:    args.RootId,
```

Comment (rfoyard) [reply server/command_hooks.go]: I already tried this, but then, clicking the button will trigger an error in `http_hooks.go`: `createPost: Invalid RootId parameter`. And if I remove the `RootId`, the grouping is not respected anymore.

Comment (hanzei) [reply server/command_hooks.go]: Could you please double check? My code change is about the post created via the slash command, not when the button is pressed.

Comment (rfoyard) [reply server/command_hooks.go]: I know, but in `http_hooks.go` I use the `request.PostId` as the new Post's RootId and when I change your line, it fails...

Comment (hanzei) [reply server/command_hooks.go]: > I know, but in `http_hooks.go` I use the `request.PostId` as the new Post's RootId and when I change your line, it fails...

I see. But this is actually not correct. If the post with the buttons is not a root post, we should not use it's id as RootId. Instead we should get the post with `request.PostId` and use **it's** rootId.

Comment (rfoyard) [reply server/command_hooks.go]: Thanks, it fixed the issue !

File: server/http_hooks.go [hunk 7]
```diff
@@ -229,6 +231,51 @@ func (p *Plugin) handleEphemeralDelete(w http.ResponseWriter, r *http.Request) {
...
+		p.API.LogError("failed to marshal json for interactive action", "err", jsonErr.Error())
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	msg := "@%v clicked an interactive button.\n```json\n%v\n```"
+	if _, appErr := p.API.CreatePost(&model.Post{
+		UserId:    p.botId,
+		ChannelId: request.ChannelId,
+		RootId:    post.RootId,
```
Comment (hanzei) [inline server/http_hooks.go]: If `post.RootId` is `""`, we need to use `post.Id` here.

**Comment by hanzei** [comment]
@prapti Please take another look

**Comment by hanzei** [comment]
@rfoyard Can you please merge `master` into your branch? The changes to `go.*` are not longer needed.

**Comment by hanzei** [comment]
Thanks for this great addition to the demo plugin :+1:

**Comment by rfoyard** [comment]
> Thanks for this great addition to the demo plugin 👍
You're very welcome !

**Pull request merged**